azure-iot-hub-devicesdk/azure/iot/hub/devicesdk/transport/mqtt/mqtt_provider.py

```diff
@@ -49,21 +50,37 @@ def _create_mqtt_client(self):
         def on_connect_callback(client, userdata, flags, result_code):
             logger.info("connected with result code: %s", str(result_code))
             # TODO: how to do failed connection?
-            self.on_mqtt_connected("connected")
+            try:
+                self.on_mqtt_connected()
+            except:  # noqa: E722 do not use bare 'except'
+                logger.error("Unexpected error calling on_mqtt_connected")
+                logger.error(traceback.format_exc())
```

Comment:
nit: seems like a pattern worth extracting in a helper function?

Reply:
That's a good idea.  I could see value in adding some sort of context guard, like:
with log_and_swallow_all_exceptions():
    self.on_mqtt_disconnected("disconnected")

with the log_and_swallow guard doing the try/except big and the logging.  

I think this is possible, but context guards might not be the right way to do it.  I suspect a decorator might be better.  But I don't want to fall into the "learning a new python language feature" trap right now.


---
In reply to: [237658858](https://github.com/Azure/azure-iot-sdk-python-preview/pull/26#discussion_r237658858) [](ancestors = 237658858)